internal/action/base.go

```diff
@@ -0,0 +1,40 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package action
```

Comment:
I'm thinking about packages and names, and I'd like to discuss it a little bit further: I can see an action as an intrinsic part of the deployment operator, for that I'd see it with good eyes to include the ServiceOperation in the `deploy` package.

Besides that, because this is very similar to the Command pattern, I'd rename the `Attach` method to `Run()`, `Exec()` or `Execute()`, with no parameters. All the input parameters could be part of the state of the disposable ServiceOperation as attributes in the struct on creation.